internal/grpctest/tlogger.go

```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpctest
+
+import (
+	"testing"
+)
+
+var logger = tLogger{infoLog, 2, &testing.T{}}
+
+type logLevel int
+
+// tLogger is a logger used for testing that logs using the testing package's T type.
+type tLogger struct {
+	l logLevel
+	v int
+	t *testing.T
+}
+
+const (
+	infoLog logLevel = iota
+	warningLog
+	errorLog
+	fatalLog
+)
+
+func (g *tLogger) setLevel(l logLevel) {
+	g.l = l
+}
+
+func (g *tLogger) setVerbosity(v int) {
+	g.v = v
+}
+
+func (g *tLogger) logIf(l logLevel, args ...interface{}) {
```

Comment:
Also, I think we'll want to call `t.Helper()` when logging so that the file/line number printed is accurate.

...and now I see that this will be invoked indirectly (`grpclog.Log->logger.Log`) and `Helper` doesn't accept a parameter to say that it should ignore other calling stack frames, so it will just end up printing the file/line in `grpclog` that called this function instead.

One *possible* way to make that work right is to have `grpclog.Log`/etc as pointers that are set when setting the logger:

```go
package grpclog

var Info func(...interface{})

func SetLoggerV2(l LoggerV2) {
  Info = l.Info
  ...
}
```

This would remove the trampoline through the other function, but it would also make the function pointer settable by end-users.  (We could move things to internal or just document them to restrict others from doing that.)  CC @menghanl -- any other ideas?

Reply:
This is not a new problem for test logger. The line numbers were already messed up for transports because of [this wrapper](https://github.com/grpc/grpc-go/blob/660df6a06bb124baedd2e36fbae01d74616d9e28/internal/transport/log.go#L30).

Maybe we need a `grpclog.SkipFrame()`? `grpclog.Helper()`?
This would only for the `t.Helper()` style functions, not for [`printDepth()`](https://github.com/golang/glog/blob/23def4e6c14b4da8ac2ed8007337bc5eb5007998/glog.go#L640)

Reply:
> Maybe we need a grpclog.SkipFrame()? grpclog.Helper()?
> This would only for the t.Helper() style functions, not for printDepth()

How would those things work though?  `Helper` works by adding the current function's name to a map, and then skipping over that function name in the backtrace when logging the file/line info.

Since `t` needs to do the logging for us, that isn't possible here.  Unless you're suggesting we accept a `*testing.T` somewhere in the logging package and call `Helper` on it in all our logging functions, but that doesn't sound right to me.  Or if grpc is generating the file/line number then that would work, but it would result in redundant (and also incorrect) info when logging via `t` or `glogger`.

There are two ways we can do this to make it so we directly call into the logger from the logging site, which also fixes this problem:
1. Replace the logging functions with pointers to the logger's implementation, e.g. `grpclog.Info = myLogger.Info` (in `SetLoggerV2`)
2. Provide access to the current logger (via `internal`) so that logging sites invoke it directly, e.g. `internal.logger.Info(...)`.

This would break anyone (including our `glogger`) assuming something about our stack frame depth, but I don't think that should be considered part of our API.

Otherwise, I think we're looking at something like #289, if we want grpc to be generating the file & line number information.